src/core/src/core_logic/PatchInstaller.py

```diff
@@ -177,14 +179,48 @@ def install_updates(self, maintenance_window, package_manager, simulate=False):
 
+        batch_patching_stopwatch = Stopwatch(self.env_layer, self.telemetry_writer, self.composite_logger)
+        batch_patching_stopwatch.start()
+
+        installed_update_count, patch_installation_successful, maintenance_window_batch_cutoff_reached = self.install_patches_in_batches(all_packages, packages, package_versions, maintenance_window, package_manager)
+
+        batch_patching_stopwatch.stop()
+
+        # Check for updates available to install. After installing patches in batches, there might be still some packages available for installation. These packages should be:
+        # (a) Due to not enough remaining time in maintenance window to install the updates in batch, the batch patching is stopped and hence some updates are not installed in batches
+        # (b) Those which were failed while installing in batches
+        # (c) Some new packages available in this brief time when batch patching was being done
+        packages, package_versions = self.get_remaining_packages_to_install(package_manager)
+
+        batch_processing_perf_log  = {"Installed patches count in batch processing": str(installed_update_count), "number of remaining packages to install": str(len(packages)), 
+                                      Constants.PerfLogTrackerParams.PATCH_OPERATION_SUCCESSFUL: str(patch_installation_successful), "Stopped batch patching due to not enough remaining time to install in batches": str(maintenance_window_batch_cutoff_reached)}
+
+        batch_patching_stopwatch.write_telemetry_for_stopwatch(str(batch_processing_perf_log))
+
+        if len(packages) == 0:
+            installed_update_count = self.log_metrics_and_perform_final_reconciliation(package_manager, maintenance_window, patch_installation_successful, False, installed_update_count)
+            return installed_update_count, patch_installation_successful, False
+        else:
+            progress_status = self.progress_template.format(str(datetime.timedelta(minutes=maintenance_window.get_remaining_time_in_minutes())), str(self.attempted_parent_update_count), str(self.successful_parent_update_count), str(self.failed_parent_update_count), str(installed_update_count - self.successful_parent_update_count),
+                                                        "Following packages are not installed in batch patching: " + str(packages))
+            self.composite_logger.log(progress_status)
+
         for package, version in zip(packages, package_versions):
+            if package not in self.last_still_needed_packages:
+                self.composite_logger.log("The following package is already installed, it should have been installed as dependent package of some other package: " + package)
+                self.attempted_parent_update_count += 1
+                self.successful_parent_update_count += 1
+                continue
+
+            per_patch_install_stopwatch = Stopwatch(self.env_layer, self.telemetry_writer, self.composite_logger)
```

Comment:
rename per_patch_install_stopwatch  to single_<>_install_stopwatch where <> is update or package, whichever is finalized in the earlier part of the code

Reply:
Updated to single_package_install_stopwatch